pkg/sql/schemachanger/corpus/corpus.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package corpus
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scop"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// Collector used to collect declarative schema changer into a corpus file.
+type Collector struct {
```

Comment:
Nice work! Could you add some commentary on how the collector works from high level? specifically, how schema changer states are cached and merged in memory through the tests, why do we need the file lock, and how do we handle corpus generated from different logic test configs (sorry, for this, I think my question is that would the parent test name be the same across different configs and there'd be conflict on corpus file names on a same test).